testdata/generate_test_data/create_complex_circular_references.py

```diff
@@ -0,0 +1,203 @@
+from pathlib import Path
+from typing import Any
+
+from lxml import etree
+
+from dsp_tools import excel2xml
+
+
+def create_and_save_circular_references_test_graph(replication_counter: int = 1, save_location: str = "") -> None:
+    """
+    This function creates a graph with circular references.
+    By default, it saves it in the current directory; this can be changed with the parameter save_location.
+
+    Args:
+        replication_counter: number of times the sub-graphs should be created in one root-graph.
+        save_location: path to the folder, where the file should be saved
+    """
+    root = create_circular_references_test_graph(replication_counter=replication_counter)
+    excel2xml.write_xml(root, Path(save_location) / f"test_circular_references_{replication_counter}.xml")
+
+
+def create_circular_references_test_graph(replication_counter: int) -> etree._Element:
+    """
+    This function creates a graph with circular references.
+    It is capable of reproducing one graph with all the references a specified number of times.
+    Each circle type has its own sub-graph number in order to differentiate them by their ID.
+    The resulting etree is suitable for an upload to the DSP-API.
+
+    Args:
+        replication_counter: number of times the sub-graphs should be created in one root-graph.
+
+    Returns:
+        A etree which is suitable for an upload into the DSP-API
+    """
+    root = excel2xml.make_root("0700", "simcir")
+    for i in range(1, replication_counter + 1):
+        root.extend(_make_one_circle_with_three_resources(replication_counter=f"{i}1"))
+        root.extend(_make_complex_dependencies(replication_counter=f"{i}2"))
+        root.extend(_make_complex_dependencies_add_on(replication_counter=f"{i}3"))
+        root.extend(_make_two_references(replication_counter=f"{i}4"))
+        root.extend(_make_chain(replication_counter=f"{i}5"))
+        root.extend(_make_inverted_complex_dependencies(replication_counter=f"{i}6"))
+    return root
+
+
+def _make_list_of_resources(number_of_resources: int, replication_counter: str, start_range: int = 65) -> list[Any]:
+    letters = [chr(x) for x in range(start_range, start_range + number_of_resources)]
+    return [
+        excel2xml.make_resource(
+            restype=":TestThing", label=f"res_{let}_{replication_counter}", id=f"res_{let}_{replication_counter}"
+        )
+        for let in letters
+    ]
+
+
+def _make_salsah_link(target_res: etree._Element) -> str:
+    return f'<a class="salsah-link" href="IRI:{target_res.attrib["id"]}:IRI">{target_res.attrib["id"]}</a>'
```

Comment:
One rule I like to follow is to try to provide a function with only as much information as it really needs. Therefor I would rather pass a string `target_id` than an entire XML element, of which you only use a single attribute.

Reply:
Not only Balduin likes to follow it, but it even has a name: Law of Demeter, a.k.a. principle of least knowledge 😄

Reply:
The reason I did this was because accessing the ID needs a bit of unpacking which I only have to do once instead of having to type this every time. But I do also see your point that this limits the usability.